Add fullscreen and app crop selection options

karaggeorge opened the pull request:
This PR adds the following features: https://github.com/wulkano/kap/issues/300, https://github.com/wulkano/kap/issues/298, https://github.com/wulkano/kap/issues/212

In order to get this to work, I made the following package: https://github.com/karaggeorge/mac-windows that returns the windows in the current screen along with their information and given an application name it brings it to the front

I wasn't sure what the design should look like. For now, I just added another select next to the aspect ratio one. 

When an Application is selected, if a user either resizes the crop window or changes the width/height inputs the select will unselect the application. Selecting an Application will also set the appropriate width/height/aspect-ratio in the other inputs.

The package I made only allows the main window of an Application to come to the front. For example, if you have 3 different Finder windows open, only the last used one (main) will show up. The AppKit Swift library only supports that for now.

<img width="349" alt="screen shot 2017-12-12 at 16 03 27" src="https://user-images.githubusercontent.com/8822835/33908323-084c60d6-df56-11e7-9485-626e09b83b4d.png">


Gifs
--
### Recording an App:
![kapture 2017-12-12 at 13 26 27](https://user-images.githubusercontent.com/8822835/33901935-5f5b75e8-df41-11e7-9d06-ca44208a047e.gif)

### Recording Fullscreen:
![kapture 2017-12-12 at 13 29 31](https://user-images.githubusercontent.com/8822835/33901998-94591b74-df41-11e7-9cf3-c430218ec853.gif)

### Changing dimensions unselects Application
![kapture 2017-12-12 at 13 30 41](https://user-images.githubusercontent.com/8822835/33902024-ad25ebaa-df41-11e7-9c71-67e19e65be68.gif)

### Refreshing open applications
![kapture 2017-12-12 at 15 50 46](https://user-images.githubusercontent.com/8822835/33908040-1df7f306-df55-11e7-8e6f-391918b8d7a0.gif)


And the Gifs created by the first two:
--
![kapture 2017-12-12 at 13 25 50](https://user-images.githubusercontent.com/8822835/33902043-ba6b8ff4-df41-11e7-8d00-5b2a10d16cff.gif)

![kapture 2017-12-12 at 13 28 55](https://user-images.githubusercontent.com/8822835/33902049-c1545dd2-df41-11e7-9641-e50c8b49de7b.gif)

doot0 commented:
![](https://media.giphy.com/media/KI9oNS4JBemyI/giphy.gif)

This is _a m a z i n g_ work, @karaggeorge, especially as a first-time contributor. Targeting specific windows has been a want of ours for a long and we've not had the bandwidth to implement, so THANK YOU! <3

Provided the team's happy with changes then I'm absolutely down for merging this.

onordbo commented:
![](https://media.giphy.com/media/26gsspfbt1HfVQ9va/giphy.gif)

karaggeorge commented:
I would hold up on the merge for a little bit. I'm still trying to fix some small issues (like when recording full screen you can still see the cropper window). If people could download the branch and test it that would be great. It would help catch minor bugs before the merge

albinekb commented:
I think it should refresh the list on click, show a spinner if it takes more than 500ms to do so and then trannsition to a menu? Really nice!!

karaggeorge commented:
@albinekb At the moment, it refreshes everytime the application is shown. It replaces the `Select...` choice with `Loading...` while the list is being loaded. I just thought that on-click would be too often, and you can't really open/close a new application without blurring from Kap (which would cause it to hide and reload the list again on show). Thoughts?

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -406,6 +485,8 @@ document.addEventListener('DOMContentLoaded', () => {
 
   ipcRenderer.on('log', (event, msgs) => console.log(...msgs));
 
+  ipcRenderer.on('load-applications', loadApplications);
```
`load-applications` => `load-apps`

karaggeorge replied:
451f413

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
+
+  function loadApplications() {
+    // Remove existing applications
+    appSelector.querySelectorAll('option:not([disabled])').forEach(option => {
```
Use a `for-of` loop.

karaggeorge replied:
5332f59

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
+  appSelector.addEventListener('change', handleAppChange);
+
+  function loadApplications() {
+    // Remove existing applications
```
`applications` => `apps`

karaggeorge replied:
451f413

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
+    });
+
+    appSelector.querySelector('option[disabled]').text = 'Loading...';
+    appSelector.querySelector('option[disabled]').selected = true;
```
You're using `appSelector.querySelector('option[disabled]')` 4 times. Save it to a variable.

karaggeorge replied:
ea47eb8

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
+    appSelector.querySelector('option[disabled]').selected = true;
+
+    // Load applications
+    getWindows().then(windows => {
```
Use `await`

karaggeorge replied:
ea47eb8

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
+    // Load applications
+    getWindows().then(windows => {
+      appSelector.querySelector('option[disabled]').text = 'Select...';
+      windows.forEach(window => {
```
Use a `for-of` loop.

karaggeorge replied:
5332f59

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -370,9 +440,18 @@ document.addEventListener('DOMContentLoaded', () => {
 
+  ipcRenderer.on('cropper-window-opened', (event, bounds) => {
+    recordBtn.classList.add('is-cropping');
+    recordBtn.attributes['data-state'].value = 'ready-to-record';
```
This should use `recordBtn.dataset`.

karaggeorge replied:
1ce2c08

sindresorhus commented on app/src/renderer/views/main.pug:
```diff
@@ -29,6 +29,10 @@ block body
+        +select-custom.has-icon.c-select--large.app-select
+          +icon('ic_crop_free_24px').o-icon--small
+          select.app-selector(tabindex="1")
+            option(disabled=true selected=true) Loading...
```
`Loading...` => `Loading…`

karaggeorge replied:
451f413

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
+      }
+    });
+
+    appSelector.querySelector('option[disabled]').text = 'Loading...';
```
`Loading...` => `Loading…`

karaggeorge replied:
451f413

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
+
+    // Load applications
+    getWindows().then(windows => {
+      appSelector.querySelector('option[disabled]').text = 'Select...';
```
`Select...` => `Select…`

karaggeorge replied:
451f413

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
+    return option;
+  };
+
+  const appData = {
```
This should be a `Map`, not object.

karaggeorge replied:
211e9ca

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +45,70 @@ document.addEventListener('DOMContentLoaded', () => {
 
   const [micOnIcon, micOffIcon] = toggleAudioRecordBtn.children;
 
+  const createOption = (label, disabled = false) => {
```
`disabled` here creates a [boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap) when used. Use an options object instead.

karaggeorge replied:
45c39dd

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -14,6 +15,8 @@ const {app} = remote;
 // Observers that should be disposed when the window unloads
 const observersToDispose = [];
 
+const cropperWindowBuffer = 2;
```
It's not clear enough from the name what this is and why it's 2. Needs a succinct code comment.

karaggeorge replied:
I got that from [here](https://github.com/wulkano/kap/blob/master/app/src/main/index.js#L32). I'm not 100% sure what its purpose is. I'm guessing to make up for the 1px of the border of the cropper window. Is that correct?

sindresorhus replied:
It was originally added in https://github.com/wulkano/kap/commit/13a0cd6d66666cd2f5f395ebf001a1de4ced50d1

karaggeorge replied:
ed705f7

sindresorhus commented on app/src/main/index.js:
```diff
@@ -173,6 +177,25 @@ ipcMain.on('open-cropper-window', (event, size) => {
+    cropperWindow.close();
+  }
+
+  activateWindow(appName).then(() => {
```
Use `await`

karaggeorge replied:
ea47eb8

sindresorhus commented on app/src/renderer/css/main.css:
```diff
@@ -239,6 +239,10 @@ body.is-tray-arrow-hidden {
 
 
 /* Aspect Ratio Controls */
```
`Aspect ratio controls`

karaggeorge replied:
451f413

sindresorhus commented:
Generally looks very good. I just commented about some nitpick :)

karaggeorge commented:
@sindresorhus 
Good catches! I fixed everything except for that `cropperWindowBuffer` variable, whose use I'm not 100% sure about.

sallar commented:
@karaggeorge The `mac-windows` package throws an error. Both in this PR and also when installed standalone on the machine:

```
{ Error: Command failed: /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows true
dyld: Symbol not found: __TMSS
  Referenced from: /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows
  Expected in: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx/libswiftCore.dylib
 in /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows

    at ChildProcess.exithandler (child_process.js:272:12)
    at ChildProcess.emit (events.js:159:13)
    at maybeClose (internal/child_process.js:943:16)
    at Socket.stream.socket.on (internal/child_process.js:363:11)
    at Socket.emit (events.js:159:13)
    at Pipe._handle.close [as _onclose] (net.js:568:12)
  killed: false,
  code: null,
  signal: 'SIGABRT',
  cmd: '/Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows true' }
(node:10153) UnhandledPromiseRejectionWarning: TypeError: windows.filter is not a function
    at getWindows.then.then.windows (/Users/sallar/tmp/node_modules/mac-windows/index.js:41:22)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:160:7)
(node:10153) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:10153) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

karaggeorge commented:
@sallar Thanks for catching that. I installed it standalone and it worked on my machine with the published build, so that's a bit weird. (it obviously also worked in the app)

I updated the package (to make sure if an exception is thrown it's handled), re-built it and updated the dependency here. Can you check again?

sallar commented:
@karaggeorge still no dice... running the binary throws this for me. I wonder if something on my machine is breaking it. but it shouldn't be because its a standalone binary right?

```
✔︎ tmp /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows true
dyld: Symbol not found: __TMSS
  Referenced from: /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows
  Expected in: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx/libswiftCore.dylib
 in /Users/sallar/tmp/node_modules/mac-windows/swift/MacWindows/.build/debug/MacWindows
fish: '/Users/sallar/tmp/node_modules/…' terminated by signal SIGABRT (Abort)
```

karaggeorge commented:
@sallar seems like other people have been having that issue:

>Oh, that's right. You can't (well, shouldn't) use binaries compiled with one version of Swift compiler in a runtime with another version.

My guess is that you have a different swift version than I do. I'll check to see what the best solution for this would be. Pretty new to swift, so I'm not quite sure.

sallar commented:
@karaggeorge yes thats my guess too. I have swift 4. But what if users dont have swift or have another version of swift installed?

albinekb commented:
@sallar the swift should be compiled before published to npm. Look at wulkano/aperture 👍

karaggeorge commented:
@albinekb I did compile before I published it to npm, but I had Swift 3, so my compiled (with Swift 3) script couldn't run on a computer running Swift 4. 

@sallar, since this app only targets Mac users, I think it's a given a swift script will run on their computer (it's compiled into an executable). However, like you said what if they have a different version. I just upgraded to swift 4 and I'll try rebuilding now. I hope the Swift 4 compiled script will work for Swift 3. Otherwise maybe keeping both versions and checking the user's current environment?

sallar commented:
@karaggeorge Thanks! Could you please do that rebuild and also fix this `main.js` conflict so we'll be able to merge?
@sindresorhus could you please check the PR again to see if we can approve for merging?

I need to do some work based on this PR and that can't be done until this is merged :)

Thanks all 🙌

matheuss commented:
> so my compiled (with Swift 3) script couldn't run on a computer running Swift 4.

This doesn’t sound right - if you compile anything into a binary, you don’t shouldn’t need any of the build-time environment if you static link everything. I don’t remember having issues like that on aperture 🤔

sallar commented:
Thats what I mean. Maybe this time release like this: `swift build -c release` maybe that helps. The build you’re currently doing is “debug” which might be the reason why it wont work.

karaggeorge commented:
Like I said before, it's my first time working with Swift, so I'm don't really know how all of this works that well. I'll try building it as a release and let you know 👍

sindresorhus commented:
@karaggeorge You need to build it with a statically bundled version of the Swift runtime as Swift is not yet ABI stable. Use this for building: `swift build --configuration=release -Xswiftc -static-stdlib`.

karaggeorge commented:
@sindresorhus that did the trick 👍 

@sallar can you give it another try?

sallar commented:
@karaggeorge Thanks! That did the trick and it now works. If you could also resolve that conflict so @sindresorhus can check again would be nice. Thanks!

karaggeorge commented:
@sallar should be all good now!

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +44,70 @@ document.addEventListener('DOMContentLoaded', () => {
+      height -= app.kap.cropperWindowBuffer;
+      ipcRenderer.send('open-cropper-window', {width, height}, {x: 1, y: 1});
+    } else {
+      ipcRenderer.send('activate-application', appName, appData.get(appName));
```
`application` => `app`

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +44,70 @@ document.addEventListener('DOMContentLoaded', () => {
+  appSelector.addEventListener('change', handleAppChange);
+
+  async function loadApps() {
+    // Remove existing applications
```
apps

sindresorhus commented on app/src/renderer/js/main.js:
```diff
@@ -41,6 +44,70 @@ document.addEventListener('DOMContentLoaded', () => {
+    disabledAppOption.text = 'Loading…';
+    disabledAppOption.selected = true;
+
+    // Load applications
```
apps

sindresorhus commented on app/package.json:
```diff
@@ -33,6 +33,7 @@
     "insight": "^0.8.3",
     "jquery": "^3.2.1",
     "lodash": "^4.17.4",
+    "mac-windows": "^0.5.4",
```
You need to install it with Yarn, so it's added to the lockfile. Just run `yarn`.

sindresorhus commented:
@karaggeorge This looks very good. Just some nitpick feedback left:

- When clicking the Kap menubar icon it shows "Select…" in the dropdown, then quickly for some milliseconds shows "Loading…", then back to "Select…" again.
- Would be nice to include icons in the list of app windows.
- When "Fullscreen" is selected in the dropdown, you can see that the text is cropped off too early, long before the down arrow icon. 
<img width="156" alt="screenshot 2018-01-04 at 00 41 39" src="https://user-images.githubusercontent.com/170270/34544785-b7865c5c-f0e8-11e7-97f7-0424d1eeb13d.png">
- When choosing "Fullscreen", the recording misses a few pixels on the edges of the screen.
<img width="409" alt="screen shot 2018-01-04 at 00 46 46" src="https://user-images.githubusercontent.com/170270/34544801-cbe34ce6-f0e8-11e7-9958-827ea9a41b88.png">

The left is my screen, the right is the recording. I intentionally moved the window some pixel from the edge to show that the white stripe is not in the recording.

sallar commented:
@sindresorhus @karaggeorge Im working on a PR that adds those icons and some other improvements.

sindresorhus commented:
Ok, good to know. Ignore that here then :)

sindresorhus commented:
Merging to unblock @sallar.

sindresorhus commented:
Really nice work @karaggeorge 🙌
